### PR TITLE
1271: Webrev generation should be optional

### PR DESCRIPTION
**Pull request by zhaosongzs**
Currently, webrevs are generated automatically when a user pushes a new commit, and it's mandatory to choose either JSON or HTML when configuring the mlbridge bot. If both of them are disabled in the configuration, the bot will throw errors.

In this patch, generating a webrev will be optional. If both JSON and HTML are disabled, no webrev will be generated.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace

### Issue
 * [SKARA-1271](https://bugs.openjdk.org/browse/SKARA-1271): Webrev generation should be optional

### Reviewers
 * [Erik Joelsson](https://openjdk.org/census#erikj) (@erikj79 - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/skara pull/1475/head:pull/1475` \
`$ git checkout pull/1475`

Update a local copy of the PR: \
`$ git checkout pull/1475` \
`$ git pull https://git.openjdk.org/skara pull/1475/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 1475`

View PR using the GUI difftool: \
`$ git pr show -t 1475`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/skara/pull/1475.diff">https://git.openjdk.org/skara/pull/1475.diff</a>

</details>

### CODE REVIEW

**Comment by zhaosongzs** [comment]
Above `ArchiveMessages#composeConversationFooter`, there is a comment says `// When changing this, ensure that the PR pattern in the notifier still matches`. I tried to figure out where the PR pattern is, but I couldn't find it.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java [hunk 1]
```diff
@@ -315,7 +315,7 @@ static String composeConversationFooter(PullRequest pr, URI issueProject, String
                 "Commit messages:\n" +
```
Comment (erikj79) [inline bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:315]:
```
// When changing this, ensure that the PR pattern in the notifier still matches
```
Have you verified this?

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java [hunk 2]
```diff
@@ -328,18 +328,22 @@ static String composeMergeConversationFooter(PullRequest pr, Repository localRep
...
-            webrevLinks = "The webrev" + (webrevs.size() > 1 ? "s" : "") + " contain" + (webrevs.size() == 1 ? "s" : "") + " " +
-                    (containsConflicts ? "the conflicts with " + pr.targetRef() : "") +
-                    (containsConflicts && containsMergeDiffs ? " and " : "") +
-                    (containsMergeDiffs ? "the adjustments done while merging with regards to each parent branch" : "")
-                    +":\n" +
-                    webrevs.stream()
-                           .map(d -> String.format(" - %s: %s", d.shortLabel(), d.uri()))
-                           .collect(Collectors.joining("\n")) + "\n\n";
+            if (webrevs.stream().noneMatch(w -> w.uri() != null)) {
+                webrevLinks = "Webrev is disabled\n\n";
```
Comment (erikj79) [inline bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:332]: I think this should just be `""` or possibly `"\n"`, not sure.

**Review by erikj79 — COMMENTED** [review]
This looks like it solves the stated problem, but the WebrevStorage is still initiated and will still do some work, like creating directories for webrev creation. Ideally I would like that whole part of the bot be optional, but it seems really hard to achieve without having to add null checks all over the place.

I'm not sure what the best solution is here. Maybe we can introduce another `WebrevDescription.Type.NONE` and if neither kind of webrev has been configured, then `WebrevGenerator::generate` would return a `Webrevdescription` of `NONE` type. 

Another possible solution would be to have an `enabled()` method on `WebrevStorage` and wrap all uses of `webrevGenerator` in a check for that. That may get messy though, but would definitely reduce unnecessary work when webrevs are disabled.

Have you looked at `updateWebrevComment`. When webrevs are disabled, we shouldn't post that comment either.

**Comment by erikj79** [comment]
> Above `ArchiveMessages#composeConversationFooter`, there is a comment says `// When changing this, ensure that the PR pattern in the notifier still matches`. I tried to figure out where the PR pattern is, but I couldn't find it.

I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.

**Comment by zhaosongzs** [comment]
> This looks like it solves the stated problem, but the WebrevStorage is still initiated and will still do some work, like creating directories for webrev creation. Ideally I would like that whole part of the bot be optional, but it seems really hard to achieve without having to add null checks all over the place.
> 
> I'm not sure what the best solution is here. Maybe we can introduce another `WebrevDescription.Type.NONE` and if neither kind of webrev has been configured, then `WebrevGenerator::generate` would return a `Webrevdescription` of `NONE` type.
> 
> Another possible solution would be to have an `enabled()` method on `WebrevStorage` and wrap all uses of `webrevGenerator` in a check for that. That may get messy though, but would definitely reduce unnecessary work when webrevs are disabled.
> 
> Have you looked at `updateWebrevComment`. When webrevs are disabled, we shouldn't post that comment either.

Yesterday, I was also thinking about what's the best way to solve this problem. And I thought my current change would have the least impact to current code. But your opinion is right, it still trigger some unnecessary work. I would think more about it.
And yes, I already made change in `updateWebrevComment`, it won't post comment in pr when webrev is disabled.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java [hunk 3]
```diff
@@ -315,7 +315,7 @@ static String composeConversationFooter(PullRequest pr, URI issueProject, String
                 "Commit messages:\n" +
                 formatCommitMessagesBrief(commits, commitsLink).orElse("") + "\n\n" +
                 "Changes: " + pr.changeUrl() + "\n" +
-                " Webrev: " + webrev.uri().toString() + "\n" +
+                (webrev.uri() == null ? "" : " Webrev: " + webrev.uri().toString()) + "\n" +
```
[the remark this answers is not in the document]
Comment (zhaosongzs) [reply bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:318]: Sure, will fix it.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java (hunk 1 above)
Comment (zhaosongzs) [reply bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:315]: I tried to find the related PR pattern again this morning, but unfortunately I still couldn't find it. I'm still confused about where to look for it. AFAIK, it's just the footer of the email.  Is there some place in the code where we would use this email and and parse it? I also tried resetting my head to the commit where the comment was added, but I couldn't find anything related to the notifier.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java (hunk 2 above)
Comment (zhaosongzs) [reply bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:332]:
```
else {
            webrevLinks = "The merge commit only contains trivial merges, so no merge-specific webrevs have been generated.\n\n";
        }
```
I just tried to align with this.

**Comment by zhaosongzs** [comment]
> > Above `ArchiveMessages#composeConversationFooter`, there is a comment says `// When changing this, ensure that the PR pattern in the notifier still matches`. I tried to figure out where the PR pattern is, but I couldn't find it.
> 
> I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.

Just see this comment. I will check it again.

File: bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java (hunk 2 above)
Comment (erikj79) [reply bots/mlbridge/src/main/java/org/openjdk/skara/bots/mlbridge/ArchiveMessages.java:332]: Yes, I figured that, but there is an important difference. That message explains why there is no webrevs where the user would expect them to be, but for a repo where they are completely turned off, I think we shouldn't even mention them.

**Comment by zhaosongzs** [comment]
> > Above `ArchiveMessages#composeConversationFooter`, there is a comment says `// When changing this, ensure that the PR pattern in the notifier still matches`. I tried to figure out where the PR pattern is, but I couldn't find it.
> 
> I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.

I couldn't find where this logic is. Could you please point it out to me?

**Comment by zhaosongzs** [comment]
> > > Above `ArchiveMessages#composeConversationFooter`, there is a comment says `// When changing this, ensure that the PR pattern in the notifier still matches`. I tried to figure out where the PR pattern is, but I couldn't find it.
> > 
> > 
> > I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.
> 
> I couldn't find where this logic is. Could you please point it out to me?

I think I found it...

**Comment by zhaosongzs** [comment]
> > Above `ArchiveMessages#composeConversationFooter`, there is a comment says `// When changing this, ensure that the PR pattern in the notifier still matches`. I tried to figure out where the PR pattern is, but I couldn't find it.
> 
> I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.

The logic you mentioned seems like has been deleted in SKARA-386. So I think we are safe.

**Comment by erikj79** [comment]
> > I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.
> 
> The logic you mentioned seems like has been deleted in SKARA-386. So I think we are safe.

The logic I was thinking about is in `MailingListArchiveReaderBot` `PULL_REQUEST_LINK_PATTERN`.

**Comment by zhaosongzs** [comment]
> > > I believe it's the logic that parses email from the email list server and extracts the PR URL from the verify first email in the each thread. You are likely safe here as you aren't changing the line that contains the PR URL, but better be safe.
> > 
> > 
> > The logic you mentioned seems like has been deleted in SKARA-386. So I think we are safe.
> 
> The logic I was thinking about is in `MailingListArchiveReaderBot` `PULL_REQUEST_LINK_PATTERN`.

You're right, it was not deleted, just moved.

**Comment by zhaosongzs** [comment]
> This looks like it solves the stated problem, but the WebrevStorage is still initiated and will still do some work, like creating directories for webrev creation. Ideally I would like that whole part of the bot be optional, but it seems really hard to achieve without having to add null checks all over the place.
> 
> I'm not sure what the best solution is here. Maybe we can introduce another `WebrevDescription.Type.NONE` and if neither kind of webrev has been configured, then `WebrevGenerator::generate` would return a `Webrevdescription` of `NONE` type.
> 
> Another possible solution would be to have an `enabled()` method on `WebrevStorage` and wrap all uses of `webrevGenerator` in a check for that. That may get messy though, but would definitely reduce unnecessary work when webrevs are disabled.
> 
> Have you looked at `updateWebrevComment`. When webrevs are disabled, we shouldn't post that comment either.

Checked the logic again and I found although `WebrevStorage` is still initiated, it will just generate some paths but not create directories.

**Review by erikj79 — APPROVED** [review]

**Comment by zhaosongzs** [comment]
Thanks for the review!
/integrate

**Pull request closed without merging**